common/factorization/par_ilu_kernels.hpp.inc

```diff
@@ -33,6 +33,203 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+    {
+        const auto length = static_cast<IndexType>(last - first);
+        const auto pos =
+            group_wide_search(IndexType{}, length, subwarp_grp,
+                              [&](IndexType i) { return first[i] >= value; });
+        return pos < length && first[pos] == value;
+    }
+};
+
+
+}  // namespace detail
+
+
+// SubwarpSize needs to be a power of 2
+// Each subwarp works on one row
+template <bool IsSorted, int SubwarpSize, typename IndexType>
+__global__
+    __launch_bounds__(default_block_size) void find_missing_diagonal_elements(
+        IndexType num_rows, IndexType num_cols,
+        const IndexType *__restrict__ col_idxs,
+        const IndexType *__restrict__ row_ptrs,
+        IndexType *__restrict__ elements_to_add_per_row,
+        bool *__restrict__ changes_required)
+{
+    const auto total_thread_count =
+        static_cast<size_type>(blockDim.x) * gridDim.x / SubwarpSize;
+    const auto tidx =
+        threadIdx.x + static_cast<size_type>(blockIdx.x) * blockDim.x;
+    const auto begin_row = static_cast<IndexType>(tidx / SubwarpSize);
+
+    auto thread_block = group::this_thread_block();
+    auto subwarp_grp = group::tiled_partition<SubwarpSize>(thread_block);
+    const auto subwarp_idx = subwarp_grp.thread_rank();
+
+    bool local_change{false};
+    for (IndexType row = begin_row; row < num_rows; row += total_thread_count) {
+        if (row >= num_cols) {
+            if (subwarp_idx == 0) {
+                elements_to_add_per_row[row] = 0;
+            }
+            continue;
+        }
+        const auto *start_cols = col_idxs + row_ptrs[row];
+        const auto *end_cols = col_idxs + row_ptrs[row + 1];
+        if (detail::find_helper<IsSorted>::find(subwarp_grp, start_cols,
+                                                end_cols, row)) {
+            if (subwarp_idx == 0) {
+                elements_to_add_per_row[row] = 0;
+            }
+        } else {
+            if (subwarp_idx == 0) {
+                elements_to_add_per_row[row] = 1;
+            }
+            local_change = true;
+        }
+    }
+    // Could also be reduced (not sure if that leads to a performance benefit)
+    if (local_change && subwarp_idx == 0) {
+        *changes_required = true;
+    }
+}
+
+
+// SubwarpSize needs to be a power of 2
+// Each subwarp works on one row
+template <int SubwarpSize, typename ValueType, typename IndexType>
+__global__
+    __launch_bounds__(default_block_size) void add_missing_diagonal_elements(
+        IndexType num_rows, const ValueType *__restrict__ old_values,
+        const IndexType *__restrict__ old_col_idxs,
+        const IndexType *__restrict__ old_row_ptrs,
+        ValueType *__restrict__ new_values,
+        IndexType *__restrict__ new_col_idxs,
+        const IndexType *__restrict__ row_ptrs_addition)
+{
+    // Precaution in case not enough threads were created
+    const auto total_thread_count =
```

Comment:
I added it because I kind of remembered that HIP had issues with `gridDim.x` overflowing.
I just talked with @yhmtsai and apparently, each dimension is stored in a 32 bit integer (meaning `blockDim.x` has 10 bit, while `gridDim.x` has 22 bit).
I guess I can remove it, if you want. In theory, this allows us to also support huge matrices without needing to use the X or Z dimension, but so far, we never needed it.